### PR TITLE
Fix notification to myself for Tickets; fixes #4643

### LINKED ISSUE #4643: [9.3.1] "Notifications for my changes" overriden by mail follow-up
* Steps to reproduce (which actions have you made) :
Select "No" in "Notifications for my changes"

* Expected result :
Expect to not be notified about your own follow-ups in a ticket

* Actual result :
If the follow-up is done by mail, you do receive the notification

* Your GLPI setup (you can find it in Setup > General menu, System tab) :
```
 Operating system: Linux glpi 4.9.0-6-amd64 #1 SMP Debian 4.9.88-1+deb9u1 (2018-05-07) x86_64
PHP 5.6.35 cgi-fcgi (Core, PDO, Reflection, SPL, SimpleXML, Zend OPcache, apc, apcu, cgi-fcgi, ctype, curl, date, dom, ereg,
	fileinfo, filter, gd, gettext, hash, iconv, imap, json, ldap, libxml, mbstring, mcrypt, mysql, mysqli, mysqlnd, pcre, pdo_pgsql,
	pgsql, posix, readline, session, snmp, soap, standard, tokenizer, xml, xmlrpc, xmlwriter, zlib)
Setup: max_execution_time="30" memory_limit="128M" post_max_size="8M" safe_mode="" session.save_handler="files"
	upload_max_filesize="10M" 
Software: lighttpd/1.4.48
	Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/69.0.3497.92 Safari/537.36
Server Software: MariaDB Server
	Server Version: 10.1.32-MariaDB
```

### PR DESCRIPTION
**Pull request by cedric-anne**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #4643 #335

### CODE REVIEW

File: inc/ticket.class.php [hunk 1]
```diff
@@ -2572,7 +2572,7 @@ function updateDateMod($ID, $no_stat_computation = false, $users_id_lastupdater
 
       if ($this->getFromDB($ID)) {
          if (!$no_stat_computation && ($this->canTakeIntoAccount() || isCommandLine())) {
-            return $this->update(
```
Comment (cedric-anne) [inline inc/ticket.class.php]: The `return` was preventing the execution of ` parent::updateDateMod()`.

File: inc/notificationtarget.class.php [hunk 2]
```diff
@@ -184,9 +186,15 @@ function getFromDBForTarget($notifications_id, $type, $ID) {
    function validateSendTo($event, array $infos, $notify_me = false) {
 
       if (!$notify_me) {
-         if (isset($infos['users_id'])
-            // Check login user and not event launch by crontask
-             && ($infos['users_id'] === Session::getLoginUserID(false))) {
+         $last_updater_id = null;
```
Comment (yllen) [inline inc/notificationtarget.class.php]: When the mailcollector work by cron, you don't have a session.

Comment (cedric-anne) [reply inc/notificationtarget.class.php]: Yes, it is exactly why I introduced ability to get the last updater id from the object, and not using the current session.

Comment (yllen) [reply inc/notificationtarget.class.php]: And how do you know that the last updater is the author of the mail?
You can have many people working in the same ticket.

Comment (cedric-anne) [reply inc/notificationtarget.class.php]: I guess "last_updater_id" field corresponds to the last updater of the ticket.
But the problem is not here. Some notifications could be made to warn users about the state of an object, independently from its update (For example warn user that a ticket he creates was not taken into accouint after a defined period). In this case, notification have to be sent to last updater.

So the only valid solution I see for the moment is to change the signature of `NotificationEvent::raiseEvent()`, `NotificationEventInterface::raise()` and `NotificationTarget::validateSendTo()` to add an optionnal parameter `$trigerrer_user_id` that can be used to force identification of the user that triggers the notification.

**Comment by cedric-anne** [comment]
Solution is not correct. I am closing this PR.

**Pull request closed without merging**